### PR TITLE
RFC6717 conformancy

### PR DESCRIPTION
**Pull request by Uzlopak**
Changes:
removing basic-auth-package
Sending by default the auth-token charset="UTF-8".
Checking for control characters in user-id and password (lacking in basic-auth-package)
adding a strictCredentials options as basic-auth implemented actually the credentials schema wrong. BWS is not in the ABNF schema of the credentials schema. BWS is extended to also cover tab.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -120,6 +120,22 @@ fastify.setErrorHandler(function (err, req, reply) {
...
 ## Options
 
+### `utf8` <Boolean> (optional, default: true)
+
+User-ids or passwords containing characters outside the US-ASCII
+character repertoire will cause interoperability issues, unless both
+communication partners agree on what character encoding scheme is to
+be used. If utf8 is set to true the server will send the 'charset' parameter
+to indicate a preference of "UTF-8", increasing the probability that
+clients will switch to that encoding.
```
Comment (jsumners) [inline README.md:130]:
```suggestion
User-ids or passwords containing characters outside the US-ASCII
character set will cause interoperability issues, unless both
communication partners agree on what character encoding scheme is to
be used. If utf8 is set to true the server will send the 'charset' parameter
to indicate a preference of "UTF-8", increasing the probability that
clients will switch to that encoding.
```

Comment (Uzlopak) [reply README.md:130]: I took that sentence from the RFC :). But i like your suggestion better.

Comment (jsumners) [reply README.md]: The person who wrote the RFC needs to crack open a dictionary https://www.merriam-webster.com/dictionary/repertoire

File: index.js [hunk 2]
```diff
@@ -56,24 +140,30 @@ async function fastifyBasicAuth (fastify, opts) {
   }
 }
 
-function getAuthenticateHeader (authenticate) {
-  if (!authenticate) return false
+function getAuthenticateHeader (authenticate, useUtf8) {
```
Comment (climba03003) [inline index.js:143]: I don't think it is a good idea to create infinite amount of function.

Comment (Uzlopak) [reply index.js:143]: Its not creating an infinite amount of functions. The function generated by getAuthenticateHeader is assigned once in line 76
https://github.com/fastify/fastify-basic-auth/blob/889cc95787755ae4077b2c4c10a5aa30c4ddb102/index.js#L76
and then reused by the plugin instance.

**Comment by Uzlopak** [comment]
Based on using autocannon on the example.js I can determine that the performance did not degrade on invalid and valid credentials.

File: README.md [hunk 3]
```diff
@@ -120,6 +120,22 @@ fastify.setErrorHandler(function (err, req, reply) {
...
+### `utf8` <Boolean> (optional, default: true)
+
+User-ids or passwords containing characters outside the US-ASCII
+character set will cause interoperability issues, unless both
+communication partners agree on what character encoding scheme is to
+be used. If utf8 is set to true the server will send the 'charset' parameter
+to indicate a preference of "UTF-8", increasing the probability that
+clients will switch to that encoding.
+
+### `strict` <Boolean> (optional, default: true)
```
Comment (Uzlopak) [inline README.md:132]: Maybe instead of strict call it strictHeader or something like that? Please give your opinion ;)

File: README.md [hunk 4]
```diff
@@ -120,6 +120,23 @@ fastify.setErrorHandler(function (err, req, reply) {
...
+be used. If utf8 is set to true the server will send the 'charset' parameter
+to indicate a preference of "UTF-8", increasing the probability that
+clients will switch to that encoding.
+
+### `strictCredentials` <Boolean> (optional, default: true)
+
+If strictCredentials is set to false the authorization header can contain
+additional whitespaces at the beginning, in the midde and at the end of the
+authorization header.
+This is a fallback option to ensure the same behaviour as @fastify/basic-auth
```
Comment (Eomm) [inline README.md:137]:
```suggestion
This is a fallback option to ensure the same behaviour as `@fastify/basic-auth`
```

**Comment by Uzlopak** [comment]
I could set strictCredentials default to false, if you worried about releasing a major after a short time?

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**